marlowe-playground-client/src/Component/MetadataTab/Types.purs

```diff
@@ -1,40 +1,111 @@
 module Component.MetadataTab.Types where
 
+import Prologue
+import Contrib.Halogen.State.Record (Update)
+import Contrib.Type.Proxy.Generic (fromRowLabels) as Proxies
+import Contrib.Variant (tag) as Contrib.Variant
+import Data.Map.Ordered.OMap (OMap)
+import Data.Newtype (class Newtype, un)
+import Data.Tuple.Nested (type (/\))
+import Data.Variant (Variant)
+import Data.Variant (inj) as Variant
 import Marlowe.Extended (ContractType)
 import Marlowe.Extended.Metadata (NumberFormat)
-import Marlowe.Semantics as S
+import Marlowe.Semantics (TokenName) as S
+import Prim.Row (class Cons) as Row
+import Type.Prelude (class IsSymbol, Proxy(..))
 
-class ShowConstructor a where
-  showConstructor :: a -> String
+-- | Let's leave this action type as it is so
+-- | it can be an example of a `Variant` based action type
+-- | and usage of different generic and non generic handlers
+-- | for `Record` based state.
+-- |
+-- | There are three types of actions here:
+-- |
+-- | - `set*` - handled generically. Actions which have a clear semantic and just
+-- | set a particular value in a state record.
+-- |
+-- | - `update*` - handled generically. Actions which have probably to broad semantic
+-- | and can modify a particular field value. In this particular case we could possibly
+-- | replace all actions with just a bunch of `update*` and move handlers logic
+-- | entirely to template. I'm just not sure if such a move is a good idea.
+-- |
+-- | - Others. I've left these "custom" actions to show how to
+-- | handle them in our final state handler together with generic handlers.
+-- |
+-- |
+-- | `Variant` usage notes:
+-- |
+-- | * We close the row here and wrap the type in a `newtype` so actually we
+-- | are building a closed sum type here. We do this because there is no
+-- | obvious benefit of keeping the row polymorphic. In theory if we would have
+-- | multiple components with shared subset of actions then possibly we could
+-- | reuse some templates / handlers etc. This scenario seems a bit unrealistic.
```

Comment:
Well, actually in MarloweRun I've seen some possible uses of this scenario. At some point we have two sets of data in the same place, contract step information and tabs UI information. They are together because we wanted to persist tabs state for different steps, but that made the logic to be tied to a particular type instead of being able to say something that has tab information and something else.

Reply:
I would like to discuss this in more details. It would be nice to check if `Variant` is going simplify the code in this other case or make it more complex :-P